NPE in machine actuator

/kind bug

**What steps did you take and what happened:**
The controller manager outputs the following NPE:
```
I1018 20:54:00.288034       1 controller.go:89] Running reconcile Cluster for test1
I1018 20:54:00.288486       1 controller.go:124] reconciling cluster object test1 triggers idempotent reconcile.
I1018 20:54:00.288494       1 actuator.go:64] Reconciling cluster test1.
I1018 20:54:00.288164       1 controller.go:113] Running reconcile Machine for aws-controlplane-g9nvr
I1018 20:54:00.288858       1 actuator.go:301] Checking if machine aws-controlplane-g9nvr for cluster test1 exists
E1018 20:54:00.289144       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go:90
/go/src/sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go:303
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/cluster-api/pkg/controller/machine/controller.go:157
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:207
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xd1a3e7]

goroutine 110 [running]:
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x1330320, 0x2005a20)
	/usr/local/go/src/runtime/panic.go:502 +0x229
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine.(*Actuator).machineProviderStatus(0xc420588750, 0xc4203f6840, 0xc4201b59b0, 0x2, 0x2)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go:90 +0x47
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine.(*Actuator).Exists(0xc420588750, 0xc4200c4d00, 0xc4203f6840, 0x0, 0x0, 0xc420509a01)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go:303 +0x112
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/cluster-api/pkg/controller/machine.(*ReconcileMachine).Reconcile(0xc4205aa580, 0xc4204b5770, 0x7, 0xc420509aa0, 0x16, 0x3, 0x3, 0x3, 0x0)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/cluster-api/pkg/controller/machine/controller.go:157 +0x397
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc4204ccb40, 0x0)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:207 +0x139
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x36
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4200219e0)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4200219e0, 0x3b9aca00, 0x0, 0x1, 0xc42059e360)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4200219e0, 0x3b9aca00, 0xc42059e360)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:156 +0x388
```


**What did you expect to happen:**
No NPE

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

--- a/Makefile
+++ b/Makefile
@@ -28,15 +28,11 @@ endif
 
 # Dependency managemnt.
 vendor:
-	dep version || go get -u github.com/golang/dep/cmd/dep
+	${GOPATH}/bin/dep version || go get -u github.com/golang/dep/cmd/dep
 	${GOPATH}/bin/dep ensure
 
 check-install:
 	./scripts/check-install.sh
-
-depend-update:
-	dep version || go get -u github.com/golang/dep/cmd/dep
-	${GOPATH}/bin/dep ensure -update
 
 # Generate code.
 generate: vendor

--- a/cmd/clusterawsadm/client/client.go
+++ b/cmd/clusterawsadm/client/client.go
@@ -57,13 +57,16 @@ func homeDir() string {
 
 func MachineInstanceID(name string) (string, error) {
 	c := NewClient()
+
 	m, err := c.ClusterV1alpha1().Machines("default").Get(name, v1.GetOptions{})
 	if err != nil {
 		return "", err
 	}
-	status, err := providerv1.MachineStatusFromBytes(m.Status.ProviderStatus.Raw)
+
+	status, err := providerv1.MachineStatusFromProviderStatus(m.Status.ProviderStatus)
 	if err != nil {
 		return "", err
 	}
+
 	return *status.InstanceID, nil
 }

--- a/cmd/manager/main.go
+++ b/cmd/manager/main.go
@@ -20,6 +20,9 @@ import (
 	"flag"
 	"os"
 
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/cluster"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine"
 	clusterapis "sigs.k8s.io/cluster-api/pkg/apis"
 	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
 	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
@@ -29,10 +32,6 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
-
-	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis"
-	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/cluster"
-	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine"
 )
 
 func main() {
@@ -54,6 +53,7 @@ func main() {
 	if err != nil {
 		panic(err)
 	}
+
 	clusterActuator, _ := cluster.NewActuator(cluster.ActuatorParams{
 		ClustersGetter: cs.ClusterV1alpha1(),
 	})
@@ -64,13 +64,17 @@ func main() {
 
 	// Register our cluster deployer (the interface is in clusterctl and we define the Deployer interface on the actuator)
 	common.RegisterClusterProvisioner("aws", clusterActuator)
+
 	if err := apis.AddToScheme(mgr.GetScheme()); err != nil {
 		panic(err)
 	}
+
 	if err := clusterapis.AddToScheme(mgr.GetScheme()); err != nil {
 		panic(err)
 	}
+
 	capimachine.AddWithActuator(mgr, machineActuator)
+
 	capicluster.AddWithActuator(mgr, clusterActuator)
 
 	if err := mgr.Start(signals.SetupSignalHandler()); err != nil {

--- a/pkg/apis/awsprovider/v1alpha1/register.go
+++ b/pkg/apis/awsprovider/v1alpha1/register.go
@@ -52,11 +52,13 @@ func ClusterStatusFromProviderStatus(extension *runtime.RawExtension) (*AWSClust
 	if extension == nil {
 		return &AWSClusterProviderStatus{}, nil
 	}
-	var status AWSClusterProviderStatus
-	if err := yaml.Unmarshal(extension.Raw, &status); err != nil {
+
+	status := new(AWSClusterProviderStatus)
+	if err := yaml.Unmarshal(extension.Raw, status); err != nil {
 		return nil, err
 	}
-	return &status, nil
+
+	return status, nil
 }
 
 func MachineConfigFromProviderConfig(providerConfig clusterv1.ProviderConfig) (*AWSMachineProviderConfig, error) {
@@ -67,10 +69,15 @@ func MachineConfigFromProviderConfig(providerConfig clusterv1.ProviderConfig) (*
 	return &config, nil
 }
 
-func MachineStatusFromBytes(raw []byte) (*AWSMachineProviderStatus, error) {
-	var status AWSMachineProviderStatus
-	if err := yaml.Unmarshal(raw, &status); err != nil {
+func MachineStatusFromProviderStatus(extension *runtime.RawExtension) (*AWSMachineProviderStatus, error) {
+	if extension == nil {
+		return &AWSMachineProviderStatus{}, nil
+	}
+
+	status := new(AWSMachineProviderStatus)
+	if err := yaml.Unmarshal(extension.Raw, status); err != nil {
 		return nil, err
 	}
-	return &status, nil
+
+	return status, nil
 }

--- a/pkg/cloud/aws/actuators/machine/actuator.go
+++ b/pkg/cloud/aws/actuators/machine/actuator.go
@@ -24,7 +24,6 @@ import (
 	"github.com/aws/aws-sdk-go/service/elb"
 	"github.com/golang/glog"
 	"github.com/pkg/errors"
-	"k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsprovider/v1alpha1"
 	service "sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services"
 	ec2svc "sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/ec2"
@@ -34,24 +33,14 @@ import (
 	controllerError "sigs.k8s.io/cluster-api/pkg/controller/error"
 )
 
-// codec are the functions off the generated codec that this actuator uses.
-type codec interface {
-	DecodeFromProviderConfig(clusterv1.ProviderConfig, runtime.Object) error
-	DecodeProviderStatus(*runtime.RawExtension, runtime.Object) error
-	EncodeProviderStatus(runtime.Object) (*runtime.RawExtension, error)
-}
-
 // Actuator is responsible for performing machine reconciliation.
 type Actuator struct {
-	codec codec
-
 	machinesGetter client.MachinesGetter
 	servicesGetter service.Getter
 }
 
 // ActuatorParams holds parameter information for Actuator.
 type ActuatorParams struct {
-	Codec          codec
 	MachinesGetter client.MachinesGetter
 	ServicesGetter service.Getter
 }
@@ -59,7 +48,6 @@ type ActuatorParams struct {
 // NewActuator returns an actuator.
 func NewActuator(params ActuatorParams) (*Actuator, error) {
 	res := &Actuator{
-		codec:          params.Codec,
 		machinesGetter: params.MachinesGetter,
 		servicesGetter: params.ServicesGetter,
 	}
@@ -80,27 +68,19 @@ func (a *Actuator) elb(clusterConfig *v1alpha1.AWSClusterProviderConfig) service
 }
 
 func (a *Actuator) machineProviderConfig(machine *clusterv1.Machine) (*v1alpha1.AWSMachineProviderConfig, error) {
-	machineProviderCfg := &v1alpha1.AWSMachineProviderConfig{}
-	err := a.codec.DecodeFromProviderConfig(machine.Spec.ProviderConfig, machineProviderCfg)
-	return machineProviderCfg, err
+	return v1alpha1.MachineConfigFromProviderConfig(machine.Spec.ProviderConfig)
 }
 
 func (a *Actuator) machineProviderStatus(machine *clusterv1.Machine) (*v1alpha1.AWSMachineProviderStatus, error) {
-	status := &v1alpha1.AWSMachineProviderStatus{}
-	err := a.codec.DecodeProviderStatus(machine.Status.ProviderStatus, status)
-	return status, err
+	return v1alpha1.MachineStatusFromProviderStatus(machine.Status.ProviderStatus)
 }
 
 func (a *Actuator) clusterProviderStatus(cluster *clusterv1.Cluster) (*v1alpha1.AWSClusterProviderStatus, error) {
-	providerStatus := &v1alpha1.AWSClusterProviderStatus{}
-	err := a.codec.DecodeProviderStatus(cluster.Status.ProviderStatus, providerStatus)
-	return providerStatus, err
+	return v1alpha1.ClusterStatusFromProviderStatus(cluster.Status.ProviderStatus)
 }
 
 func (a *Actuator) clusterProviderConfig(cluster *clusterv1.Cluster) (*v1alpha1.AWSClusterProviderConfig, error) {
-	providerConfig := &v1alpha1.AWSClusterProviderConfig{}
-	err := a.codec.DecodeFromProviderConfig(cluster.Spec.ProviderConfig, providerConfig)
-	return providerConfig, err
+	return v1alpha1.ClusterConfigFromProviderConfig(cluster.Spec.ProviderConfig)
 }
 
 // Create creates a machine and is invoked by the machine controller.
@@ -126,12 +106,6 @@ func (a *Actuator) Create(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	if err != nil {
 		return errors.Wrap(err, "failed to get machine config")
 	}
-
-	defer func() {
-		if err := a.updateStatus(machine, status); err != nil {
-			glog.Errorf("failed to store provider status for machine %q: %v", machine.Name, err)
-		}
-	}()
 
 	i, err := a.ec2(clusterConfig).CreateOrGetMachine(machine, status, config, clusterStatus, cluster)
 	if err != nil {
@@ -288,11 +262,6 @@ func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 		}
 	}
 
-	// Finally update the machine status.
-	if err := a.updateStatus(machine, status); err != nil {
-		return errors.Wrap(err, "failed to update machine status")
-	}
-
 	return nil
 }
 
@@ -347,24 +316,6 @@ func (a *Actuator) updateMachine(machine *clusterv1.Machine) error {
 
 	if _, err := machinesClient.Update(machine); err != nil {
 		return fmt.Errorf("failed to update machine: %v", err)
-	}
-
-	return nil
-}
-
-func (a *Actuator) updateStatus(machine *clusterv1.Machine, status *v1alpha1.AWSMachineProviderStatus) error {
-	machinesClient := a.machinesGetter.Machines(machine.Namespace)
-
-	encodedProviderStatus, err := a.codec.EncodeProviderStatus(status)
-	if err != nil {
-		return fmt.Errorf("failed to encode machine status: %v", err)
-	}
-
-	if encodedProviderStatus != nil {
-		machine.Status.ProviderStatus = encodedProviderStatus
-		if _, err := machinesClient.UpdateStatus(machine); err != nil {
-			return fmt.Errorf("failed to update machine status: %v", err)
-		}
 	}
 
 	return nil

--- a/pkg/cloud/aws/services/ec2/instances.go
+++ b/pkg/cloud/aws/services/ec2/instances.go
@@ -17,14 +17,12 @@ import (
 	"encoding/base64"
 	"fmt"
 
-	"github.com/golang/glog"
-
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
 	"github.com/pkg/errors"
-	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
-
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsprovider/v1alpha1"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
 // InstanceByTags returns the existing instance or nothing if it doesn't exist.

